### ISSUE TITLE
pyarrow write_data_set max_open_files not closing files

### ISSUE THREAD

**Issue by davlee1972** [apache/arrow#38427]
### Describe the bug, including details regarding any error messages, version, and platform.

We have ulimit(s) set in linux so I tried using the @max_open_file parameter in pyarrow.Dataset.write_dataset().

My python program just hangs and it seems like the write_dataset() function is not closing files so it can't open new ones..
I can see the amount of files written appears to be exactly double when @max_open_files is set to double at the point where everything hangs..

I'm also writing out partioned data in the following directory structure:
parted_parquet/end_date_year=2019/end_date_month=9/end_date_day=9/flows.0.parquet

with max_open_files = 256
du -h -s parted_parquet
1.3G    parted_parquet

with max_open_files = 128
du -h -s parted_parquet
686M    parted_parquet

### Component(s)

Python

**Comment by mapleFU** [comment]
> My python program just hangs and it seems like the write_dataset() function is not closing files so it can't open new ones

This might slightly related to the implemention of DatasetWriter, let me dive into it...

**Comment by zackmedi** [comment]
I'm also running in to this issue currently on Python 3.11.6, pyarrow 14.0.1

**Comment by mapleFU** [comment]
@zackmedi Do you have some reproduce steps for this? I'm trying to reading code here.

**Comment by mapleFU** [comment]
https://github.com/apache/arrow/pull/38885/files

This might be related

**Comment by zackmedi** [comment]
@mapleFU Here's a script for reproducing the issue!

Something I just noticed while creating this script is that it seems to depend on my `ulimit -n` setting on my terminal. As you can see below the `max_open_files = 20`, however the script will hang if I set `ulimit -n 20` or `ulimit -n 25`. I didn't test every individual value, but at `ulimit -n 30` the script starts working again. So a ulimit value somewhere between 5 and 10 files over the `max_open_files` arg fixes the issue in this case. I'm not sure how it scales in the case where max_open_files is at 30,000 like when I originally encountered it though.

```
import random
import shutil
import string
from pathlib import Path
import pyarrow
import pyarrow.dataset as ds

def gen_batches(my_schema):
    batch_dict = create_empty_dict()
    for i in range(1000):
        batch_dict['part'].append(i)

        chars = ''.join(random.choice(string.ascii_letters) for _ in range(3))
        batch_dict['chars'].append(chars)

        if len(batch_dict['part']) >= 3:
            yield pyarrow.RecordBatch.from_pydict(batch_dict, my_schema)
            batch_dict = create_empty_dict()

def create_empty_dict():
    return {
        'part': [],
        'chars': []
    }

if __name__ == '__main__':
    shutil.rmtree('./tmp/', ignore_errors=True)
    Path('./tmp/').mkdir(exist_ok=True)

    my_schema = pyarrow.schema([
        ('part', pyarrow.uint32()),
        ('chars', pyarrow.string())
    ])

    writer_args = {
        'schema': my_schema,
        'format': 'parquet',
        'partitioning': [
            'part'
        ],
        'partitioning_flavor': 'hive',
        'max_partitions': 1_000,
        'max_open_files': 20,
        'existing_data_behavior': 'delete_matching'
    }

    ds.write_dataset(
        gen_batches(my_schema),
        base_dir='./tmp/',
        **writer_args
    )
```

**Comment by zackmedi** [comment]
@mapleFU I'm still having issues with this with ulimit out of the way. For instance setting ulimit -n 100000 (so my OS doesn't kill the program first) and passing `max_open_files=500` to write_dataset will still hang.

**Comment by mapleFU** [comment]
Would you mind try apply this patch https://github.com/apache/arrow/pull/38885 ?

I'm not sure this patch actually fix the problem, but it seems it will cause some hang problem here.

**Comment by mapleFU** [comment]
Hmm we can using this: https://arrow.apache.org/docs/developers/python.html

But it's a bit complex

**Comment by zackmedi** [comment]
It wasn't too bad, the instructions are really well written! Unfortunately it doesn't look like it fixed the issue for me though. My script still hangs indefinitely right before it gets to the max_open_files limit.

In this test, I had set `max_open_files=100` and the process had 99 files open when it hung (according to the process' Open Files and Ports in ActivityMonitor).

<img width="803" alt="Screenshot 2023-12-04 at 11 19 51 AM" src="https://github.com/apache/arrow/assets/113271096/59eff4d4-cde1-4abb-bd07-a66513a69569">

**Comment by zackmedi** [comment]
@mapleFU any updates or timelines on this?

**Comment by mapleFU** [comment]
I've try to fix with https://github.com/apache/arrow/pull/38885 in main branch. But I'm not sure it works. I'm quite busy in these few days, let me try this.

Also, if the process hang with https://github.com/apache/arrow/pull/38885 , would you mind pstack and get some stack tracing? It would help me find where it hang

**Comment by zackmedi** [comment]
Sorry I'm not familiar with pstack, but I imagine my reproducible steps above would help you run it right?

**Comment by mapleFU** [comment]
Hmmm currently I cannot reproducing it, let me check it later. Also 

> After debugging more, it's definitely an issue even with the ulimit out of the way. ulimit -n 100000 and max_open_files=10 will still hang when the program attempts to open file # 10.

Would you mind check the output and is there only 9 files? 🤔

**Comment by zackmedi** [comment]
That's correct, it was 1 file less than the max I had set. In my specific test I set max_open_files=100 and when it hung there were 99 parquet files open in my activity monitor
